<!DOCTYPE html>
<html >
<head>

	<meta http-equiv="cache-control" content="no-cache, no-store, must-revalidate" />
	<meta http-equiv="expires" content="0" />
	<meta http-equiv="pragma" content="no-cache" />

	<!-- Due to a bug in webcomponents, I can't use CSS with HTML imports -->
	<!-- (at least in IE) -->
	<!-- https://github.com/webcomponents/webcomponentsjs/issues/400 -->
	<link rel="stylesheet" href="bootstrap/css/bootstrap-yeti.css"/>
	<link rel="stylesheet" href="css/style.css"/>
	<link rel="stylesheet" href="angular-cookie-law/angular-cookie-law.min.css"/>

	<!-- Polyfill, required for browsers that do not support link rel="import" -->
	<script type="text/javascript" src="webcomponents/webcomponents-lite.min.js"></script>

	<link rel="import" href="html/dependencies.html"/>
	
</head>
<body>

<cookie-law-banner message="This website uses Google Analytics First Party cookies for collecting usage statistics and some user-related data including geography and IP address. We use this facility to understand our users and improve our website. In order to use the website, please indicate that you consent to this." accept-text="I consent"></cookie-law-banner>

<div class="container">

	<div ng-controller="isgWebAppCtrl">
		<h2>Orthologous Clusters of Interferon-Stimulated Genes</h2>
		<div class="row"">
			<div class="col-md-8"
			
					<p>One of the key innate immune mechanisms that antagonise
						virus infection in vertebrates is the interferon (IFN) system.
						Type I interferons (including IFN-&beta; and IFN-&alpha; among others),
						type-II IFN (IFN-&gamma;) and type III interferons (IFN-&lambda;), are
						cytokines with immunomodulatory and proinflammatory properties.
						IFN signalling results in the upregulation of hundreds of
						interferon stimulated genes (ISGs), collectively referred to as
						the interferome, that initiate an antiviral state. This database
						catalogues the interferomes of ten different species, each derived
						from cells in an antiviral state following stimulation with type-I
						IFN. In this study an ISG is a gene upregulated with a false
						discovery rate (FDR) &lt; 0.05, regardless of the extent to which it
						is upregulated. This can be adjusted in the user parameters.</p>
				</div>
			<div class="col-md-4"
				<p><a href="http://www.gla.ac.uk/researchinstitutes/iii/cvr/" target="_blank"><img alt="MRC-University of Glasgow Centre for Virus Research" src="images/CVRLogo.png"/></a></p>
				<p><a href="http://www.gla.ac.uk/researchinstitutes/iii/cvr/" target="_blank"><small>MRC-University of Glasgow Centre for Virus Research</small></a></p>
			</div>
		</div>
		<cookie-law-wait>
			<uib-accordion ng-cloak close-others="true">
			<uib-accordion-group is-open="searchByGeneOrEnsemblId">
				<uib-accordion-heading>
					Search clusters by gene symbol or ENSEMBL ID
					<i class="pull-right glyphicon" ng-class="{'glyphicon-chevron-down': searchByGeneOrEnsemblId, 'glyphicon-chevron-right': !searchByGeneOrEnsemblId}"></i>
				</uib-accordion-heading>
			
				<table class="table table-nonfluid borderless" style="margin: 0px">
					<thead>
						<colgroup>
							<col span="1" style="width: 40%;">
							<col span="1" style="width: 60%;">
						</colgroup>
				
						<tr>							
								<td>
									<autocomplete ng-model="geneOrEnsemblNameTyped" attr-input-class="geneEnsemblInput" attr-ul-class="geneEnsemblDropdown"
										data="geneOrEnsemblSuggestions" on-type="updateGeneOrEnsemblSuggestions" ng-keydown="myKeyDown($event)"
										on-select="updateSelectedGeneOrEnsembl" click-activation="true" no-auto-sort="true" attr-placeholder="Enter gene symbol* or ENSEMBL ID"></autocomplete>
								</td>
						          <td>
									<button type="button" class="btn btn-default" ng-click="runGeneOrEnsemblQuery()">Run query</button>
						          </td>
						</tr>
						<tr>
							<td>* Search using an HGNC approved gene symbol.<br/>Use the <a target="_blank" href="https://www.genenames.org/cgi-bin/symbol_checker">HGNC symbol checker</a> to look up approved gene symbols.</td>
						</tr>
							
				</table>
	        </uib-accordion-group>
			<uib-accordion-group is-open="searchByPresenceExpressionSpecific">
				<uib-accordion-heading>
					Search clusters by presence / expression in specific species
					<i class="pull-right glyphicon" ng-class="{'glyphicon-chevron-down': searchByPresenceExpressionSpecific, 'glyphicon-chevron-right': !searchByPresenceExpressionSpecific}"></i>
				</uib-accordion-heading>
				<p>Selected clusters must meet all of the following species criteria:</p>
				<table class="table table-bordered"  style="margin: 0px;">
					<tbody>
						<tr>							
							<td></td>
							<td>
									Ortholog:
								<div class="btn-group">
									<label class="btn btn-primary btn-xs ng-pristine ng-untouched ng-valid active" 
										ng-click="setOnAllSpeciesCriteria('requirePresent')">Present</label>
									<label class="btn btn-primary btn-xs ng-pristine ng-untouched ng-valid active" 
										ng-click="setOnAllSpeciesCriteria('requireAbsent')">Absent</label>
								</div>
							</td>
							<td>
								<i class="glyphicon glyphicon-arrow-up upregulated-icon"></i> Up-regulated gene:
								<div class="btn-group">
									<label class="btn btn-primary btn-xs ng-pristine ng-untouched ng-valid active" 
										ng-click="setOnAllSpeciesCriteria('requireUpregulated')">Present</label>
									<label class="btn btn-primary btn-xs ng-pristine ng-untouched ng-valid active" 
										ng-click="setOnAllSpeciesCriteria('requireNotUpregulated')">Absent</label>
								</div>
							</td>
							<td>
								<i class="glyphicon glyphicon-arrow-down downregulated-icon"></i> Down-regulated gene:
								<div class="btn-group">
									<label class="btn btn-primary btn-xs ng-pristine ng-untouched ng-valid active" 
										ng-click="setOnAllSpeciesCriteria('requireDownregulated')">Present</label>
									<label class="btn btn-primary btn-xs ng-pristine ng-untouched ng-valid active" 
										ng-click="setOnAllSpeciesCriteria('requireNotDownregulated')">Absent</label>
								</div>
							</td>
						</tr>
					
					
					
						<tr ng-repeat="speciesCriterion in speciesCriteria">							
							<td>{{species[speciesCriterion.species].displayName}}</td>
							<td>
								Ortholog:
								<div class="btn-group">
									<label class="btn btn-default btn-xs ng-pristine ng-untouched ng-valid active" 
										uib-btn-checkbox ng-model="speciesCriterion.requirePresent">Present</label>
									<label class="btn btn-default btn-xs ng-pristine ng-untouched ng-valid active" 
										uib-btn-checkbox ng-model="speciesCriterion.requireAbsent">Absent</label>
								</div>
							</td>
							<td>
								<i class="glyphicon glyphicon-arrow-up upregulated-icon"></i> Up-regulated gene:
								<div class="btn-group">
									<label class="btn btn-default btn-xs ng-pristine ng-untouched ng-valid active" 
										uib-btn-checkbox ng-model="speciesCriterion.requireUpregulated">Present</label>
									<label class="btn btn-default btn-xs ng-pristine ng-untouched ng-valid active" 
										uib-btn-checkbox ng-model="speciesCriterion.requireNotUpregulated">Absent</label>
								</div>
							</td>
							<td>
								<i class="glyphicon glyphicon-arrow-down downregulated-icon"></i> Down-regulated gene:
								<div class="btn-group">
									<label class="btn btn-default btn-xs ng-pristine ng-untouched ng-valid active" 
										uib-btn-checkbox ng-model="speciesCriterion.requireDownregulated">Present</label>
									<label class="btn btn-default btn-xs ng-pristine ng-untouched ng-valid active" 
										uib-btn-checkbox ng-model="speciesCriterion.requireNotDownregulated">Absent</label>
								</div>
							</td>
						</tr>
					</tbody>	
				</table>
				<table class="table table-nonfluid" style="margin: 0px;">
					<tr>
						<td style="border-style: none; padding-left: 0px;">
							<button type="button" class="btn btn-default"
								ng-click="runSpeciesCriteriaQuery()">Run query</button>
						</td>
						<td style="border-style: none; padding-left: 0px;">
							<button class="btn btn-default" ng-click="resetSpeciesCriteria()">Reset species criteria</button>
						</td>
					</tr>
				</table>

			</uib-accordion-group>

			<uib-accordion-group is-open="searchByPresenceExpressionNumber">
				<uib-accordion-heading>
					Search clusters by presence / expression in a number of species
					<i class="pull-right glyphicon" ng-class="{'glyphicon-chevron-down': searchByPresenceExpressionNumber, 'glyphicon-chevron-right': !searchByPresenceExpressionNumber}"></i>
				</uib-accordion-heading>
				<p>Selected clusters must meet all of the following criteria:</p>
				<table class="table table-bordered table-nonfluid"  style="margin: 0px;">
					<tbody>
						<tr>
							<td>Orthologs must be present in</td>
							<td>at least 
								<input class="param-text" type="text" ng-model="searchByNumberCriteria.presentMin"/> species and at most 
								<input class="param-text" type="text" ng-model="searchByNumberCriteria.presentMax"/> species</td>
						</tr>
						<tr>
							<td>
								<i class="glyphicon glyphicon-arrow-up upregulated-icon"></i>
								Up-regulated genes must be present in
							</td>
							<td>at least 
								<input class="param-text" type="text" ng-model="searchByNumberCriteria.upRegulatedPresentMin"/> species and at most 
								<input class="param-text" type="text" ng-model="searchByNumberCriteria.upRegulatedPresentMax"/> species</td>
						</tr>
						<tr>
							<td>
								<i class="glyphicon glyphicon-arrow-down downregulated-icon"></i>
								Down-regulated genes must be present in
							</td>
							<td>at least 
								<input class="param-text" type="text" ng-model="searchByNumberCriteria.downRegulatedPresentMin"/> species and at most 
								<input class="param-text" type="text" ng-model="searchByNumberCriteria.downRegulatedPresentMax"/> species</td>
						</tr>
					</tbody>
				</table>
				<table class="table table-nonfluid" style="margin: 0px;">
					<tr>
						<td style="border-style: none; padding-left: 0px;">
							<button type="button" class="btn btn-default"
								ng-click="runSearchByNumberQuery()">Run query</button>
						</td>
						<td style="border-style: none; padding-left: 0px;">
							<button class="btn btn-default" ng-click="resetSearchByNumberCriteria()">Reset criteria</button>
						</td>
					</tr>
				</table>
			
	        </uib-accordion-group>
			<uib-accordion-group is-open="editGeneExpressionParams">
				<uib-accordion-heading>
					Differential expression parameters
					<i class="pull-right glyphicon" ng-class="{'glyphicon-chevron-down': searchByPresenceExpression, 'glyphicon-chevron-right': !searchByPresenceExpression}"></i>
				</uib-accordion-heading>
				How genes are classified as up-regulated or down-regulated:
				<table class="table table-nonfluid" style="margin: 0px;">
					<tr>
						<td style="border-style: none; padding-left: 0px;">
							<table class="table table-bordered table-nonfluid"  style="margin: 0px;">
								<tbody>
									<tr>
										<td><i
											class="glyphicon glyphicon-arrow-up upregulated-icon"></i> <b>Up-regulated
												genes:</b></td>
										<td>log<sub>2</sub> FC &gt; <input class="param-text"
											type="text"
											ng-model="geneRegulationParams.upregulatedMinLog2FC" /> and FDR &lt; {{geneRegulationParams.maxFDR}}</td>
									</tr>
									<tr>
										<td><i
											class="glyphicon glyphicon-arrow-down downregulated-icon"></i>
											<b>Down-regulated genes:</b></td>
										<td>log<sub>2</sub> FC &lt; <input class="param-text"
											type="text"
											ng-model="geneRegulationParams.downregulatedMaxLog2FC" /> and FDR &lt; {{geneRegulationParams.maxFDR}}</td>
									</tr>
									<tr>
										<td><b>False discovery rate:</b></td>
										<td>FDR &lt; <input class="param-text" type="text"
											ng-model="geneRegulationParams.maxFDR" /></td>
									</tr>
								</tbody>
							</table>
						</td>
						<td style="border-style: none; padding-left: 0px;">
							<button class="btn btn-default"
								ng-click="resetGeneRegulationParams()">Reset paramters</button>
						</td>
					</tr>
				</table>
			</uib-accordion-group>
			<uib-accordion-group is-open="userGuide">
				<uib-accordion-heading>
					User guide
					<i class="pull-right glyphicon" ng-class="{'glyphicon-chevron-down': userGuide, 'glyphicon-chevron-right': !userGuide}"></i>
				</uib-accordion-heading>
				
				<h3>Orthologous clusters</h3>
				<p>This database groups genes together in orthologous clusters, based on ENSEMBL annotations. Clusters may contain genes from different species i.e. orthologs. 
				In some cases a gene has expanded within the species of a genome producing paralogs, so a cluster may also contain multiple genes from the same species. </p>
				
				<p>For some clusters there was no ENSEMBL-annotated ortholog in the genome of a given species. Some of these absent orthologs have been resolved through a BLAST approach, 
				and others via in silico 'baiting' for reads matching the ortholog of a closely related species. These genes are prefixed with 'CVR-defined gene'.</p>

				<p>Within an orthologous cluster, the effect of interferon varies: some genes are up-regulated, others down-regulated, still others are not differentially expressed.</p> 
				
				<h3>Searching</h3>
				This web tool allows users to search for orthlogous clusters based on various kinds of search criteria.
				
				<h5><b>Search clusters by gene symbol or ENSEMBL ID</b></h5>
				Simply type in a gene symbol or ENSEMBL ID; the web tool will suggest matches from the database and return the cluster containing the specified gene.
				Approved gene symobls provided by the <a target="_blank" href="https://www.genenames.org/">HUGO Gene Nomenclature Committee</a> (HGNC) are accepted. For example, to search for RIG-I you will need to use DDX58.
				If you are unsure of the approved gene symbol, the <a target="_blank" href="https://www.genenames.org/cgi-bin/symbol_checker">HGNC symbol checker</a> can be used to check.</td>
				
				
				<h5><b>Search clusters by presence / expression in specific species</b></h5>
				Select clusters based on the presence and expression characteristics of genes in particular species. 
				For example:
				<em>"All clusters which contain orthologs in all the mammal species, including an up-regulated gene in humans and a down-regulated gene in cattle and sheep".</em>
				
				<h5><b>Search clusters by presence / expression in a number of species</b></h5>
				Select clusters based on the presence and expression characteristics of genes in a certain number of species. 
				For example: 
				<em>"All clusters which contain an ortholog in at least 5 species, including at least 3 species where there is an up-regulated gene".</em>

				<h3>Results</h4>
				<p>All clusters matching the criteria provided will be output. Clusters will specify all the genes within the cluster, grouped by species. Some species may not posess an ortholog 
				within the cluster; these species will be absent.</p>
				<p>If a gene's differential expression value is significant, it will have a red (significantly upregulated) or blue (significantly downregulated) 
				arrow in the 'Expression' column, otherwise this column is blank.</p>
				<p>Tables of search results can be downloaded in either tab-delimited or CSV format using the 'Download' button. </p>
								
				<h3>Differential expression parameters</h3>
				This controls how genes are classified as up-regulated or down-regulated. Default values are set as any significant level of up-regulation (&gt; 0) or down-regulation (&lt; 0). 
				The default value for significance is set at a false discovery rate (FDR) of 0.05. All these parameters may be edited; changing these values can affect which clusters are 
				selected and how the 'Expression' result column is populated.

	        </uib-accordion-group>


		</uib-accordion>
		</cookie-law-wait>
			<div ng-cloak ng-show="resultRows">
				<h2>Results</h2>
				<div ng-show="resultRows.length == 0">No clusters found for this query</div>
				<div ng-show="resultRows.length > 0">
					<div class='table query-table'>
						<div class="row">
							<button type="button" class="btn btn-default" 
								ng-click="firstPage()">First</button>
							<button type="button" ng-disabled="firstClusterIndex == 1" class="btn btn-default" 
								ng-click="prevPage()">Previous</button>
							<button type="button" ng-disabled="lastClusterIndex == orthoClusters.length" class="btn btn-default" 
								ng-click="nextPage()">Next</button>
							<button type="button" class="btn btn-default" 
								ng-click="lastPage()">Last</button>
						    
							<div class="btn-group" uib-dropdown is-open="status.isopen_numclusters">
						      <button id="numclusters-button" type="button" class="btn btn-default" uib-dropdown-toggle ng-disabled="disabled">
						        Clusters per page: {{clustersPerPage}} <span class="caret"></span>
						      </button>
						      <ul class="dropdown-menu" role="menu" aria-labelledby="numclusters-button">
						        <li ng-repeat="availableClustersPerPageValue in availableClustersPerPage" ng-click="setAvailableClustersPerPage(availableClustersPerPageValue)" role="menuitem"><a >
						        	Clusters per page: {{availableClustersPerPageValue}}</a></li>
						      </ul>
							</div>
							<div class="btn-group" uib-dropdown is-open="status.isopen_download">
						      <button id="download-button" type="button" class="btn btn-default" uib-dropdown-toggle ng-disabled="disabled">
						        Download <span class="caret"></span>
						      </button>
						      <ul class="dropdown-menu" role="menu" aria-labelledby="download-button">
						        <li ng-click="downloadResults('csv')" role="menuitem"><a>Comma-separated values (CSV)</a></li>
						        <li ng-click="downloadResults('tab')" role="menuitem"><a>Tab-delimited</a></li>
						      </ul>
							</div>
							<button type="button" class="btn btn-default" 
								ng-click="clearResults()">Clear results</button>
						    
						</div>
					</div>
					<p>Clusters {{firstClusterIndex}} to {{lastClusterIndex}} of {{orthoClusters.length}}</p>
					<table class="table table-bordered tbl-results">
						<thead>
							<th>Orthologous cluster ID</th>
							<th>Species</th>
							<th>ENSEMBL ID</th>
							<th>Gene</th>
							<th>Expression</th>
							<th><em>log<sub>2</sub></em> Fold Change</em></th>
							<th>FDR</em></th>
						</thead>		
						<tbody>
							<tr class="{{resultRow.rowClass}}" ng-repeat="resultRow in resultRows track by $index">
								<td ng-show="resultRow.orthoClusterId" rowSpan="{{resultRow.numRowsInCluster}}">{{resultRow.orthoClusterId}}</td>
								<td ng-show="resultRow.species" rowSpan="{{resultRow.numRowsInSpecies}}">{{resultRow.species}}</td>
								<td ng-show="!resultRow.ensemblId.startsWith('ENSCVR')"><a href="http://www.ensembl.org/Gene/Summary?g={{resultRow.ensemblId}}" target="_blank">{{resultRow.ensemblId}}</a></td>
								<td ng-show="resultRow.ensemblId.startsWith('ENSCVR')">CVR-defined gene: {{resultRow.ensemblId}}</td>
								<td>{{resultRow.geneName}}</td>
								<td ng-show="resultRow.upregulated"><i class="glyphicon glyphicon-arrow-up upregulated-icon"></i></td>
								<td ng-show="resultRow.downregulated"><i class="glyphicon glyphicon-arrow-down downregulated-icon"></i></td>
								<td ng-show="resultRow.not_differentially_expressed"></td>
								<td>{{resultRow.log2fc}}</td>
								<td>{{resultRow.fdr}}</td>
							</tr>
						</tbody>	
					</table>
				</div>
		  </div>
	</div>
	<h3>Publication</h3>
	<div class="row"">
		<div class="col-md-3"
			<p><img alt="PLOS Biology" width="100%" src="images/plosBiologyWithStripe.png"/></p>
		</div>
		<div class="col-md-8"
			<p>
			 <h4><b>Fundamental Properties of the Mammalian Innate Immune System Revealed by 
			 	Multi-Species Comparison of Type I Interferon Responses</b></h4>
			 </p>
			<p class="small">
			 Andrew E. Shaw, Joseph Hughes, Quan Gu, Abdelkader Behdenna, Josh Singer, Tristan Dennis, 
			 Richard J. Orton, Mariana Varela, Robert J. Gifford, Sam J. Wilson and Massimo Palmarini
			</p>
			<p>
			 Published December 18, 2017 : <a href="https://doi.org/10.1371/journal.pbio.2004086">https://doi.org/10.1371/journal.pbio.2004086</a>
			</p>
		</div>
		<div class="col-md-1">
			<a href="https://twitter.com/intent/tweet" class="twitter-share-button" data-size="large" data-text="Innate immune system properties revealed by multi-species comparison of interferon responses https://doi.org/10.1371/journal.pbio.2004086 Search and download #interferon stimulated genes from the paper here: " data-hashtags="PLOSBiology,virology,innate,immunity,evolution" data-related="virusmuser,cvrinfo,SamJWilsonPhD,bljog,joshWentViral,supergu0925,rjrtn,paleovirologist,kaderbehdenna" data-show-count="false">Tweet</a><script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
		</div>
	</div>
	<h3>Feedback</h3>
	Contact <a href="http://www.gla.ac.uk/researchinstitutes/iii/staff/andrewshaw">Andrew Shaw</a> with questions about the research or <a href="http://www.gla.ac.uk/researchinstitutes/iii/staff/joshsinger/">Josh Singer</a> with any feedback on the web tool.
</div>
</body>
</html>